src/gateway/GatewayV2.sol

```diff
@@ -0,0 +1,347 @@
+            state.votes.decreaseFirstNValues(router.quorum);
+
+            // Handle message
+            if (isMessageProof) {
+                dispatch(state.pendingMessage);
+            } else {
+                dispatch(payload);
+            }
+
+            // Only if there are no more pending messages, remove the pending message
+            if (state.votes.isEmpty()) {
+                delete state.pendingMessage;
+            }
+
+            emit ExecuteMessage(payload, msg.sender);
+        } else if (!isMessageProof) {
+            state.pendingMessage = payload;
+        }
+    }
+
+    function _handleRecovery(bytes memory payload) internal {
+        if (MessagesLib.messageType(payload) == MessagesLib.Call.InitiateMessageRecovery) {
+            bytes32 messageHash = payload.toBytes32(1);
+            address router = payload.toAddress(33);
+            require(activeRouters[msg.sender].id != 0, "Gateway/invalid-sender");
+            require(activeRouters[router].id != 0, "Gateway/invalid-router");
+            recoveries[messageHash] = Recovery(block.timestamp + RECOVERY_CHALLENGE_PERIOD, router);
+            emit InitiateMessageRecovery(messageHash, router);
+        } else if (MessagesLib.messageType(payload) == MessagesLib.Call.DisputeMessageRecovery) {
+            bytes32 messageHash = payload.toBytes32(1);
+            return _disputeMessageRecovery(messageHash);
+        }
+    }
+
+    /// @inheritdoc IGatewayV2
+    function disputeMessageRecovery(bytes32 messageHash) external auth {
+        _disputeMessageRecovery(messageHash);
+    }
+
+    function _disputeMessageRecovery(bytes32 messageHash) internal {
+        delete recoveries[messageHash];
+        emit DisputeMessageRecovery(messageHash);
+    }
+
+    /// @inheritdoc IGatewayV2
+    function executeMessageRecovery(bytes calldata message) external {
+        bytes32 messageHash = keccak256(message);
+        // wouldn't it better to mark these as memory?
+        Recovery storage recovery = recoveries[messageHash];
+        Router storage router = activeRouters[recovery.router];
+
+        require(recovery.timestamp != 0, "Gateway/message-recovery-not-initiated");
+        require(recovery.timestamp <= block.timestamp, "Gateway/challenge-period-has-not-ended");
+        require(router.id != 0, "Gateway/invalid-router");
+
+        delete recoveries[messageHash];
+        _handle(message, recovery.router, router, true);
+        emit ExecuteMessageRecovery(message);
+    }
+
+    // --- Outgoing ---
+
+    /// @inheritdoc IGatewayV2
+    function send(bytes calldata message) public payable auth {
+        uint256 numRouters = routers.length;
+        require(numRouters > 0, "Gateway/not-initialized");
+        uint256 tank = fuel;
+
+        uint256 destChainMsgCost;
+        uint256 destChainProofCost;
+
+        bytes memory proof = abi.encodePacked(uint8(MessagesLib.Call.MessageProof), keccak256(message));
+
+        if (tank > 0) {
```

Comment:
there might be enough funds in the gateway for tx that didn't require funding from users (= don't need to be actively prepaid / call did not come from centrifuge router). So tank in that case can still be 0 and the tx should go through and be funded through funds held in gateway